File: back-end/Mercury.Engine/Mercury.Engine.API/Services/ChatGroupService.cs
using System.Threading.Tasks;
using Grpc.Core;
using Mercury.Engine.API.DbContext.Entity;
using Mercury.Engine.API.DbContext.UnitOfWork;
using Mercury.Engine.API.Models.GrpcGenerated;
using Mercury.Engine.API.Services.GrpcGenerated;

namespace Mercury.Engine.API.Services
{
    public class ChatGroupService : ChatGroup.ChatGroupBase
    {
        private readonly IUnitOfWork _unitOfWork;

        public ChatGroupService(IUnitOfWork unitOfWork) => _unitOfWork = unitOfWork;

        public override async Task<CreateReply> Create(CreateRequest request, ServerCallContext context)
        {
            var group = await _unitOfWork.TbGroupRepository
                .Add(new TbGroup { NmGroupName = request.GroupName })
                .ConfigureAwait(false);

            _unitOfWork.Save();

            await _unitOfWork.TbUserGroupRepository.BulkInsertAsync(request.UsersId, group.Entity.IdGroup).ConfigureAwait(false);

            _unitOfWork.Save();

            return new CreateReply { GroupId = group.Entity.IdGroup };
        }

        public override async Task GetGroupsByUser(
            GetGroupsByUserRequest request,
            IServerStreamWriter<GetGroupsByUserReply> streamWriter,
            ServerCallContext context)
        {
            var groups = _unitOfWork.TbUserGroupRepository.GetGroupsByUserId(request.UserId).ConfigureAwait(false);

            await foreach (var group in groups)
            {
                await streamWriter.WriteAsync(new GetGroupsByUserReply
                {
                    Group = new Group
                    {
                        GroupId = group.IdGroup,
                        GroupName = group.NmGroupName
                    }
                })
                .ConfigureAwait(false);
            }
        }
    }
}

File: front-end/mercury/lib/services/chat_group_service.dart
import 'package:mercury/services/service_base.dart';

import 'gen/services/chat_group.pb.dart';
import 'gen/services/chat_group.pbgrpc.dart';

class ChatGroupService extends ServiceBase {
  Future<CreateReply> requestCreate(String groupName, List<int> usersId) async {
    final clientChannel = await setup.clientChannel;
    final client = ChatGroupClient(clientChannel);

    var request = CreateRequest()
      ..groupName = groupName.trim()
      ..usersId.addAll(usersId);

    CreateReply reply;
    try {
      reply = await client.create(request);
    } catch (e) {
      print('Caught error: $e');
    }

    // await clientChannel.shutdown();
    return reply;
  }

  Stream<GetGroupsByUserReply> requestGroupsByUserId(int userId) async* {
    final clientChannel = await setup.clientChannel;
    final client = ChatGroupClient(clientChannel);

    var request = GetGroupsByUserRequest()..userId = userId;

    try {
      await for (var group in client.getGroupsByUser(request)) {
        yield group;
      }

      // client.getGroupsByUser(request).listen((z) => print(z.group.groupName));
    } catch (e) {
      print('Caught error: $e');
    }
    // await clientChannel.shutdown();
  }
}
